x-pack/legacy/plugins/code/server/queue/cancellation_service.test.ts

```diff
@@ -25,11 +25,33 @@ test('Register and cancel cancellation token', async () => {
   const cancelSpy = sinon.spy();
   token.cancel = cancelSpy;
 
-  await service.registerCancelableIndexJob(
-    repoUri,
-    token as CancellationToken,
-    Promise.resolve('resolved')
-  );
+  // make sure the promise won't be fulfilled immediately
+  const promise = new Promise(resolve => {
+    setTimeout(resolve, 100);
```

Comment:
the best way to manipulate the timer in unit test is using sinon's fake timer (example [here](https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/code/server/scheduler/update_scheduler.test.ts#L111)). Otherwise, the test will have some risks of flaky. It's up to you if you want to update it.

Reply:
The intention here is to make sure the promise is completed after `service.cancelIndexJob(repoUri)` is called, otherwise the promise will be removed from the map. I tried another way to defer its resolution and rejection, as introduced in this [question](https://stackoverflow.com/questions/26150232/resolve-javascript-promise-outside-function-scope). It seems a bit not clean here, any better ideas?

Reply:
there are some other options
1. use `Promise.deferred`, basically it's same as your current approach but the API is obsolete
2. use third party library like https://www.npmjs.com/package/delay, which is basically your previous approach

so overall, I think your current solution is good enough

Reply:
@fantapsody My point here is not the way you defer the promise wrong. I should be clear here. Even if you defer 100ms here, there are still some risks that the promise resolves before the `cancelIndexJob`. A better way is to create a faketimer and then

```
await service.registerCancelableIndexJob(repoUri, token as CancellationToken, promise);
// Tick 50ms so it's 100% sure that the promise won't complete before the next statement
faketimer.tick(50)
await service.cancelIndexJob(repoUri);
```